File: LarkCardKit/Builders/TableBuilder.cs
using LarkCardKit.Models.Elements;

namespace LarkCardKit.Builders;

/// <summary>
/// 表格构建器
/// </summary>
/// <remarks>
/// 用于构建表格组件。表格支持多种数据类型的列，包括文本、数字、选项标签、人员列表、日期等。
/// <para>注意事项：</para>
/// <list type="bullet">
///   <item><description>单张卡片最多支持放置五个表格组件</description></item>
///   <item><description>表格组件不可被内嵌在其它组件内，只可放在卡片根节点下</description></item>
///   <item><description>最多支持添加 50 列，超出 50 列的内容不展示</description></item>
/// </list>
/// </remarks>
/// <example>
/// 以下示例演示如何使用 TableBuilder：
/// <code>
/// var card = CardBuilder.Create()
///     .Header(h => h.Title("客户数据表"))
///     .Body(b => b
///         .Table(table => table
///             .PageSize(5)
///             .Column(col => col.Name("name").DisplayName("客户名称").DataType("text"))
///             .Column(col => col.Name("amount").DisplayName("金额").DataType("number"))
///             .Row(row => row.TextCell("name", "飞书科技").NumberCell("amount", 168))))
///     .Build();
/// </code>
/// </example>
public class TableBuilder
{
    private readonly Table _table = new();
    
    /// <summary>
    /// 设置元素唯一标识
    /// </summary>
    /// <param name="id">元素 ID</param>
    public TableBuilder ElementId(string id)
    {
        _table.ElementId = id;
        return this;
    }
    
    /// <summary>
    /// 设置外边距
    /// </summary>
    /// <param name="margin">外边距值</param>
    public TableBuilder Margin(string margin)
    {
        _table.Margin = margin;
        return this;
    }
    
    /// <summary>
    /// 设置每页最大展示的数据行数
    /// </summary>
    /// <param name="pageSize">行数，1-10</param>
    public TableBuilder PageSize(int pageSize)
    {
        _table.PageSize = pageSize;
        return this;
    }
    
    /// <summary>
    /// 设置表格行高
    /// </summary>
    /// <param name="rowHeight">
    /// 行高：
    /// - low：低（默认）
    /// - middle：中
    /// - high：高
    /// - auto：自适应内容
    /// - [32,124]px：自定义行高
    /// </param>
    public TableBuilder RowHeight(string rowHeight)
    {
        _table.RowHeight = rowHeight;
        return this;
    }
    
    /// <summary>
    /// 设置最大行高
    /// </summary>
    /// <param name="maxHeight">最大行高，如 50px</param>
    public TableBuilder RowMaxHeight(string maxHeight)
    {
        _table.RowMaxHeight = maxHeight;
        return this;
    }
    
    /// <summary>
    /// 设置是否冻结首列
    /// </summary>
    /// <param name="freeze">是否冻结，默认 true</param>
    public TableBuilder FreezeFirstColumn(bool freeze = true)
    {
        _table.FreezeFirstColumn = freeze;
        return this;
    }
    
    /// <summary>
    /// 设置表头样式
    /// </summary>
    /// <param name="configure">表头样式配置</param>
    public TableBuilder HeaderStyle(Action<TableHeaderStyleBuilder> configure)
    {
        var builder = new TableHeaderStyleBuilder();
        configure(builder);
        _table.HeaderStyle = builder.Build();
        return this;
    }
    
    /// <summary>
    /// 添加列定义
    /// </summary>
    /// <param name="configure">列配置</param>
    public TableBuilder Column(Action<TableColumnBuilder> configure)
    {
        var builder = new TableColumnBuilder();
        configure(builder);
        _table.Columns ??= new List<TableColumn>();
        _table.Columns.Add(builder.Build());
        return this;
    }
    
    /// <summary>
    /// 添加行数据
    /// </summary>
    /// <param name="configure">行数据配置</param>
    public TableBuilder Row(Action<TableRowBuilder> configure)
    {
        var builder = new TableRowBuilder();
        configure(builder);
        _table.Rows ??= new List<Dictionary<string, object?>>();
        _table.Rows.Add(builder.Build());
        return this;
    }
    
    /// <summary>
    /// 批量设置行数据
    /// </summary>
    /// <param name="rows">行数据列表</param>
    public TableBuilder Rows(IEnumerable<Dictionary<string, object?>> rows)
    {
        _table.Rows = rows.ToList();
        return this;
    }
    
    /// <summary>
    /// 构建表格对象
    /// </summary>
    public Table Build() => _table;
}

File: TypeScript/LarkCardKit/src/builders/TableBuilder.ts
import type { Table, TableColumn, TableHeaderStyle, TableNumberFormat } from '../models/elements';

export class TableBuilder {
  private _table: Table = { tag: 'table', columns: [], rows: [] };

  pageSize(size: number): this {
    this._table.page_size = size;
    return this;
  }

  rowHeight(height: string): this {
    (this._table as any).row_height = height;
    return this;
  }

  column(configure: (builder: TableColumnBuilder) => void): this {
    const builder = new TableColumnBuilder();
    configure(builder);
    this._table.columns ??= [];
    this._table.columns.push(builder.build());
    return this;
  }

  row(values: Record<string, any>): this {
    this._table.rows ??= [];
    this._table.rows.push(values);
    return this;
  }

  headerStyle(configure: (builder: TableHeaderStyleBuilder) => void): this {
    const builder = new TableHeaderStyleBuilder();
    configure(builder);
    this._table.header_style = builder.build();
    return this;
  }

  elementId(id: string): this {
    this._table.element_id = id;
    return this;
  }

  margin(margin: string): this {
    this._table.margin = margin;
    return this;
  }

  build(): Table {
    return this._table;
  }
}

export class TableColumnBuilder {
  private _column: TableColumn = { name: '', data_type: 'text' };

  name(name: string): this {
    this._column.name = name;
    return this;
  }

  displayName(name: string): this {
    this._column.display_name = name;
    return this;
  }

  dataType(type: string): this {
    this._column.data_type = type;
    return this;
  }

  width(width: string): this {
    this._column.width = width;
    return this;
  }

  horizontalAlign(align: string): this {
    this._column.horizontal_align = align;
    return this;
  }

  verticalAlign(align: string): this {
    this._column.vertical_align = align;
    return this;
  }

  format(configure: (builder: TableNumberFormatBuilder) => void): this {
    const builder = new TableNumberFormatBuilder();
    configure(builder);
    this._column.format = builder.build();
    return this;
  }

  build(): TableColumn {
    return this._column;
  }
}

export class TableHeaderStyleBuilder {
  private _style: TableHeaderStyle = {};

  backgroundStyle(style: string): this {
    this._style.background_style = style;
    return this;
  }

  textAlign(align: string): this {
    this._style.text_align = align;
    return this;
  }

  textSize(size: string): this {
    this._style.text_size = size;
    return this;
  }

  textColor(color: string): this {
    this._style.text_color = color;
    return this;
  }

  bold(bold: boolean = true): this {
    this._style.bold = bold;
    return this;
  }

  build(): TableHeaderStyle {
    return this._style;
  }
}

export class TableNumberFormatBuilder {
  private _format: TableNumberFormat = {};

  decimalPlaces(places: number): this {
    this._format.decimal_places = places;
    return this;
  }

  symbol(symbol: string): this {
    this._format.symbol = symbol;
    return this;
  }

  formattingType(type: string): this {
    this._format.formatting_type = type;
    return this;
  }

  build(): TableNumberFormat {
    return this._format;
  }
}
